core/trino-main/src/main/java/io/trino/execution/scheduler/EventDrivenFaultTolerantQueryScheduler.java

```diff
@@ -0,0 +1,2103 @@
+            this.taskSource = requireNonNull(taskSource, "taskSource is null");
+            this.sinkPartitioningScheme = requireNonNull(sinkPartitioningScheme, "sinkPartitioningScheme is null");
+            this.exchange = requireNonNull(exchange, "exchange is null");
+            this.partitionMemoryEstimator = requireNonNull(partitionMemoryEstimator, "partitionMemoryEstimator is null");
+            this.maxTaskExecutionAttempts = maxTaskExecutionAttempts;
+            this.schedulingPriority = schedulingPriority;
+            this.dynamicFilterService = requireNonNull(dynamicFilterService, "dynamicFilterService is null");
+            outputDataSize = new long[sinkPartitioningScheme.getPartitionCount()];
+            sinkOutputSelectorBuilder = ExchangeSourceOutputSelector.builder(ImmutableSet.of(exchange.getId()));
+            ImmutableMap.Builder<PlanFragmentId, RemoteSourceNode> remoteSources = ImmutableMap.builder();
+            ImmutableSet.Builder<PlanNodeId> remoteSourceIds = ImmutableSet.builder();
+            for (RemoteSourceNode remoteSource : stage.getFragment().getRemoteSourceNodes()) {
+                remoteSourceIds.add(remoteSource.getId());
+                remoteSource.getSourceFragmentIds().forEach(fragmentId -> remoteSources.put(fragmentId, remoteSource));
+            }
+            this.remoteSourceIds = remoteSourceIds.build();
+            this.remoteSources = remoteSources.buildOrThrow();
+        }
+
+        public StageId getStageId()
+        {
+            return stage.getStageId();
+        }
+
+        public PlanFragmentId getStageFragmentId()
+        {
+            return stage.getFragment().getId();
+        }
+
+        public StageState getState()
+        {
+            return stage.getState();
+        }
+
+        public StageInfo getStageInfo()
+        {
+            return stage.getStageInfo();
+        }
+
+        public Exchange getExchange()
+        {
+            return exchange;
+        }
+
+        public Optional<ScheduledTask> addPartition(int partitionId, NodeRequirements nodeRequirements)
+        {
+            if (getState().isDone()) {
+                return Optional.empty();
+            }
+
+            ExchangeSinkHandle exchangeSinkHandle = exchange.addSink(partitionId);
+            Session session = queryStateMachine.getSession();
+            DataSize defaultTaskMemory = stage.getFragment().getPartitioning().equals(COORDINATOR_DISTRIBUTION) ?
+                    getFaultTolerantExecutionDefaultCoordinatorTaskMemory(session) :
+                    getFaultTolerantExecutionDefaultTaskMemory(session);
+            StagePartition partition = new StagePartition(
+                    partitionId,
+                    exchangeSinkHandle,
+                    remoteSourceIds,
+                    nodeRequirements,
+                    partitionMemoryEstimator.getInitialMemoryRequirements(session, defaultTaskMemory),
```

Comment:
Good point. I honestly haven't had a chance to think about design of it more. It feels like it might be better to do this when we actually start implementing it. For simplicity i decided to avoid including memory estimates in the TaskSource interface, but there's a very good change it might end up being there.